### ISSUE TITLE
Best tools to debug a fuzzing workflow?

### ISSUE THREAD

**Issue by samstack6** [AFLplusplus/AFLplusplus#611]
Hi,

This might be more a general question rather than specific to AFL++ but I'm asking here cuz I mainly use AFL++.

Am curious to know what are the best tools to debug or troubleshoot a fuzzer on a specific target? Any specific recommendations rather than using general debuggers (gdb, ...) or rr ? any specific built-in tool specifically for AFL++ in this case?

Mostly looking forward to a debug why AFL++ is doing well on target X and why another fuzzer sucks or vice versa.

**Comment by vanhauser-thc** [comment]
this is a very good question and rather something for a thesis :) 

one approach is to use coverage (e.g. https://github.com/vanhauser-thc/afl-cov) on the discovered paths and see which paths were found and which are not, and then try to analyze where roadblocks are that the fuzzer cannot solve and what would be needed to solve it.

another approach is to use fuzzer introspection and see which kinds of mutations are responsible for finding paths (make INTROSPECTION=1), but this hurts performance and better suits fuzzer development/optimization. plus you only get this in afl++ and libfuzzer, but not in vanilla afl or honggfuzz or any other fuzzer I know.

Finally you can turn on and off various features and see if that is the reason a fuzzer is sucessful, e.g. laf-intel/compcov, redqueen/cmplog, dictionaries, etc.

basically a bad technique or even the worst idea can perform well on a specific target although it is horrible on anything else, and on the other hand something very good (our havoc strategy plus redqueen/cmplog) does not fair well against a few selected others.

Read our new README.md on how to effectively use afl++

**Comment by samstack6** [comment]
Thank you for the answer.

> one approach is to use coverage (e.g. https://github.com/vanhauser-thc/afl-cov) 

What is the difference between your version of afl-cov and the Rash's version?

> then try to analyze where roadblocks are that the fuzzer cannot solve and what would be needed to solve it

Any tools here for this step for analyzing the roadblocks? debuggers? etc?


> another approach is to use fuzzer introspection and see which kinds of mutations are responsible for finding paths (make INTROSPECTION=1), but this hurts performance and better suits fuzzer development/optimization. plus you only get this in afl++ and libfuzzer, but not in vanilla afl or honggfuzz or any other fuzzer I know.

I Didn't know ahout this. thanks to mention.

> Read our new README.md on how to effectively use afl++

I will check it thank you!

**Comment by vanhauser-thc** [comment]
> > one approach is to use coverage (e.g. https://github.com/vanhauser-thc/afl-cov)
> 
> What is the difference between your version of afl-cov and the Rash's version?

the original author does not maintain it anymore. I added a few helper scripts but more importantly:
 * afl-cov now accepts "@@" like AFL++ in the target command parameters (was AFL_FILE before that)
 * afl-cov now can send to targets that read on stdin (just omit @@)

> > then try to analyze where roadblocks are that the fuzzer cannot solve and what would be needed to solve it
> 
> Any tools here for this step for analyzing the roadblocks? debuggers? etc?

the coverage information from above can help you. you will see which paths were found/taken and which not. those which were not, analyze why

> > another approach is to use fuzzer introspection and see which kinds of mutations are responsible for finding paths (make INTROSPECTION=1), but this hurts performance and better suits fuzzer development/optimization. plus you only get this in afl++ and libfuzzer, but not in vanilla afl or honggfuzz or any other fuzzer I know.
> 
> I Didn't know ahout this. thanks to mention.

that feature is just a week old ;)
